contracts/vakro/AdelVAkroSwap.sol

```diff
@@ -0,0 +1,154 @@
+ 
+// SPDX-License-Identifier: AGPL V3.0
+pragma solidity ^0.6.12;
+
+import "@ozUpgradesV3/contracts/access/OwnableUpgradeable.sol";
+import "@ozUpgradesV3/contracts/token/ERC20/IERC20Upgradeable.sol";
+import "@ozUpgradesV3/contracts/token/ERC20/SafeERC20Upgradeable.sol";
+import "@ozUpgradesV3/contracts/math/SafeMathUpgradeable.sol";
+
+import "../../interfaces/IERC20Burnable.sol";
+import "../../interfaces/IERC20Mintable.sol";
+import "../../interfaces/delphi/IStakingPool.sol";
+
+contract AdelVAkroSwap is OwnableUpgradeable {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+    using SafeMathUpgradeable for uint256;
+ 
+    event AkroAdded(uint256 amount);
+    event AdelSwapped(address indexed receiver, uint256 adelAmount, uint256 akroAmount);
+
+    //Addresses of affected contracts
+    address public akro;
+    address public adel;
+    address public vakro;
+    address public stakingPool;
+
+    //Swap settings
+    uint256 public minAmountToSwap = 0;
+    uint256 public swapRate = 0; //Amount of vAkro for 1 ADEL
+
+    uint256 public swapLiquidity = 0;
+
+    modifier swapEnabled() {
+        require(swapRate != 0, "Swap is disabled");
+        _;
+    }
+
+    modifier enoughAdel(uint256 _adelAmount) {
+        require(_adelAmount > 0 && _adelAmount >= minAmountToSwap, "Insufficient ADEL amount");
+        _;
+    }
+
+    function initialize(address _akro, address _adel, address _vakro, address _stakingPool) public initializer {
+        require(_akro != address(0), "Zero address");
+        require(_adel != address(0), "Zero address");
+        require(_vakro != address(0), "Zero address");
+        require(_stakingPool != address(0), "Zero address");
+
+        __Ownable_init();
+
+        akro = _akro;
+        adel = _adel;
+        vakro = _vakro;
+        stakingPool = _stakingPool;
+    }    
+
+    //Setters for the swap tuning
+
+    /**
+     * @notice Sets the minimum amount of ADEL which can be swapped. 0 by default
+     * @param _minAmount Minimum amount in wei (the least decimals)
+     */
+    function setMinSwapAmount(uint256 _minAmount) public onlyOwner {
+        minAmountToSwap = _minAmount;
+    }
+
+    /**
+     * @notice Sets the rate of ADEL to vAKRO swap - how many vAKRO for 1 ADEL
+     * @notice By default is set to 0, that means that swap is disabled
+     * @param _swapRate Amout of vAKRO for 1 ADEL. Can be set to 0 - that stops the swap.
+     */
+    function setSwapRate(uint256 _swapRate) public onlyOwner {
+        swapRate = _swapRate;
+    }
+
+    /**
+     * @notice Adds AKRO liquidity to the swap contract
+     * @param _amount Amout of AKRO added to the contract.
+     */
+    function addSwapLiquidity(uint256 _amount) public {
```

Comment:
Anyone can add swap liquidity?

Reply:
Why not? Any how is willing to help the protocol. I do not see sence of extra checks for "owner-only". No money transfer except AKRO from caller to swap - it is safe from the protocol's view